Add an item-collection-tutorial guide. (#310)
cerner/terra-ui: src/terra-dev-site/guide/Component Guides/RecreatingItemCollection.guide.mdx

import ResponsiveListandTableExample from '../../example/ResponsiveListAndTableExample?dev-site-example';

> # Re-creating Item Collection

Reviewer: ```suggestion
# Building an Item Collection
```

> # Building an Item Collection

Reviewer: Rename this file to be the same as the title here. `BuildingAnItemCollection.guide.mdx`

Author: Renamed https://github.com/cerner/terra-ui/pull/310/commits/353727590506d9c58f0f44b2cbfd34a9accec9b7

The `terra-item-collection` used [terra-responsive-element](/components/terra-responsive-element/responsive-element/responsive-element) to swap between it's [terra-table](/components/terra-table/table/table) and [terra-list](/components/terra-list/list/list) displays. The example in this guide use `ActiveBreakpointContext` from [terra-application](/application/terra-application/application) to switch what is displayed based on current browser size.

```javascript
import { ActiveBreakpointContext } from 'terra-application/lib/breakpoints';
...
const CustomItemCollection = () => {
  const activeBreakpoint = React.useContext(ActiveBreakpointContext);
  ...
  switch (activeBreakpoint) {
    case 'tiny':
    case 'small':
    case 'medium':
      return listDisplay;
    case 'large':
    case 'huge':
    case 'enormous':
    default:
      return tableDisplay;
  }
};
```

> This will display our list display at browser size `medium` and below and our table display on larger sizes.

Reviewer: ```suggestion
This will display a list at browser size `medium` and below and a table on larger sizes.
```

> The example also use [terra-clinical-item-view](/components/terra-clinical-item-view/clinical-item-view/clinical-item-view) for each cell of data which is not required. It is used here to to recreate the deprecated `terra-item-collection` but you may create your own cell layout to meet the requirements of your design..

Reviewer: ```suggestion
The example also use [terra-clinical-item-view](/components/terra-clinical-item-view/clinical-item-view/clinical-item-view) for each cell of data which is not required. It is used here to to recreate the deprecated `terra-item-collection` but we encourage you to create your own cell layout to meet the requirements of your design..
```

> ## Replicating the Deprecated Terra Clinical Item Collection displays

Reviewer: ```suggestion
## Creating the Item Collection
```

> In the following examples we create 2 different displays. A [terra-list](/components/terra-list/list/list) display is used when browser size is below 'medium'. The other display is a [terra-table](/components/terra-table/table/table).

Reviewer: ```suggestion
In the following examples we create 2 different components. A [terra-list](/components/terra-list/list/list) display is used when browser size is below 'medium'. The other component is a [terra-table](/components/terra-table/table/table).
```

> ### Creating the List version

Reviewer: ```suggestion
### Creating the List
```

Start with a Terra `List` component and define these two props: `dividerStyles` and `role`. `dividerStyle` is used to determine if and where you want the border-color to be applied to each of the children and can have the values 'none', 'standard', 'bottom-only'. `role` is only used if the children you have in `List` are selectable, and if they are, is defined as 'listbox'.

```javascript
  const listDisplay = (
  <List dividerStyle="none" role="listbox">
    {children}
  </List>);
```

Next we build the children that are going into this `List`. Each child should be dropped in [`terra-list`'s `Item`](/components/terra-list/list/list-item) and be set as the display prop in an [ItemView](/components/terra-clinical-item-view/clinical-item-view/clinical-item-view).

```javascript
  const listChild = (
    <Item>
      <ItemView
        displays={React.Children.toArray(children)}
        ...
      />
    </Item>
  );
```

In our example we use many optional props to better organize and display the mock data we have provided.

> ## Creating the Table

Reviewer: ```suggestion
### Creating the Table
```

Author: fixed https://github.com/cerner/terra-ui/pull/310/commits/353727590506d9c58f0f44b2cbfd34a9accec9b7#diff-0d1ed2de83ef09cd3f3475cb7f85ec0ae3cc8018d38624c0093fffeb49af5d23R63

> ## Creating Single Selectable Table version

Reviewer: ```suggestion
## Creating the Table
```

import ResponsiveListandTableExample from '../../example/ResponsiveListAndTableExample?dev-site-example';
import ResponsiveListandCellGridExample from '../../example/ResponsiveListAndCellGridExample?dev-site-example';
import ResponsiveListandHtmlTableExample from '../../example/ResponsiveListAndHtmlTableExample?dev-site-example';

> # Creating An Item Collection

Reviewer 2: let's rename to the new title.
```suggestion
# Re-creating Item Collection
```

> An Item Collection is a wrapper component designed to display data as either a table or list of items dependent on the relative container size.

Reviewer: ```suggestion
`terra-item-collection` was a wrapper component designed to display data as either a table or list of items dependent on the relative container size. It has been deprecated in favor of encouraging teams to create their own component. We found that the item collection was too limited in API to meet the varied needs of all consumers and instead we now recommend teams use terra components to create their own flexible list/table component. This guide is intended to help consumers create their own 'item-collection' to switch between a table and list layout depending on the relative container size.
```

The `terra-item-collection` used [terra-responsive-element](/components/terra-responsive-element/responsive-element/responsive-element) to swap between it's [terra-table](/components/terra-table/table/table) and [terra-list](/components/terra-list/list/list) displays. The examples in this guide use `ActiveBreakpointContext` from [terra-application](/application/terra-application/application) to switch what is displayed based on current browser size.

```javascript
import { ActiveBreakpointContext } from 'terra-application/lib/breakpoints';
...
const CustomItemCollection = () => {
  const activeBreakpoint = React.useContext(ActiveBreakpointContext);
  ...
  switch (activeBreakpoint) {
    case 'tiny':
    case 'small':
    case 'medium':
      return listDisplay;
    case 'large':
    case 'huge':
    case 'enormous':
    default:
      return tableDisplay;
  }
};
```

This will display our list display at browser size `medium` and below and our table display on larger sizes.

> The examples also use [terra-clinical-item-view](/components/terra-clinical-item-view/clinical-item-view/clinical-item-view) for each cell of data which is not necessary and is used to to recreate the deprecated `terra-item-collection`.

Reviewer: ```suggestion
The examples also use [terra-clinical-item-view](/components/terra-clinical-item-view/clinical-item-view/clinical-item-view) for each cell of data which is not required. It is used here to to recreate the deprecated `terra-item-collection` but you may create your own cell layout to meet the requirements of your design..
```

## Replicating the Deprecated Terra Clinical Item Collection displays

> 

Reviewer: Add an introduction here talking about the three different methods of re-creating the item collection and why you'd choose one over the other.

Reviewer: Mention that the same list is used for each of the three different methods

Author: Addressed here https://github.com/cerner/terra-ui/pull/310/commits/e9da608d29aaca7a6a54bf05797065b88e72b842#diff-f306b40c35b7837e14535b95656559fca5a16191b1c2d2c2d535b3e07e66705aR35

### Creating the List version

Start with a Terra `List` component and define these two props: `dividerStyles` and `role`. `dividerStyle` is used to determine if and where you want the border-color to be applied to each of the children and can have the values 'none', 'standard', 'bottom-only'. `role` is only used if the children you have in `List` are selectable, and if they are, is defined as 'listbox'.

```javascript
  const listDisplay = (
  <List dividerStyle="none" role="listbox">
    {children}
  </List>);
```

> Next we build the children that are going into this `<List>`. Each child should be wrapped in `terra-list`'s `<Item>` and as display in an `<ItemView>` from `terra-clinical-item-view`.

Reviewer 3: Where is `children` coming from and what type are they?
```suggestion
Next we build the children that are going into this `List`. Each child should be dropped in `terra-list`'s `Item` and be set as the display prop in an `ItemView` from `terra-clinical-item-view`.
```

```javascript

>   var listChild = (

Reviewer 3: ```suggestion
  const listChild = (
```
Should this be mapping each child to a List item?

    <Item>
      <ItemView
        displays={React.Children.toArray(children)}
      />
    </Item>
  );
```

> ## Creating Single Selectable Table version

Reviewer 3: Should we start with creating the table items before implementing selectable or should the list items example above show to make the items selectable as well?

Reviewer: We need to go one way or the other, it's not acceptable to have a selectable table that transforms into a non-selectable list.

For `tableDisplay`, we start with this.

```javascript
  const tableDisplay = (
    <Table
      summaryId="standard-table"
      summary="This table has standard padding"
      numberOfColumns={mockData[0].cells.length}
      dividerStyle="horizontal"
      rowStyle="disclose"
      bodyData={[
        {

>           rows: createRows(mockData),

Reviewer 3: Should mockData be `React.Children.toArray` to align with the above example?

        },
      ]}
      className={cx('table')}
    />
  );
```

> `summaryId` is an element-id for the component, `summary` is just a brief text to describe the table's content, and `className` is just add some CSS. `numberOfColumns` is also pretty self-explanatory as how many columns you will need in the table. `dividerStyle` is used to put divisions between rows.

Reviewer 3: We should add that the className value should be a scope class name.

`rowStyle` we define as 'disclose' because in our example our rows are selectable to then disclose further information upon selection.

In our particular example, we created a mock data object we parse through to create the rows and cells. The `createRows` function which looks like this.

```javascript

> const createRow = rowData => (

Reviewer 3: Should this be looping the rowData to create the list of rows?
```suggestion
const createRows = rowData => (
```

`discloseAction` is a special object for use when we select that row. `metaData` just contains a key for that row and `discloseLabel` is just a brief description the row such as 'Power Row 0 - Details'. `isDisclosed` is 'true' only when our component's state matches with the metadata key. `discloseCellIndex` is the index of the cell that provides the accessible link entry point.

`onDisclose` is a simple state update function as seen here.

```javascript
const handleRowDisclose = (event, metaData) => {
  event.preventDefault();
  if (selectedKey !== metaData.key) {
    setSelectedKey(metaData.key);
  }
};
```

For creating cells in each row we do this

```javascript
const createCellsForRow = cells => cells.map(cell => createCell(cell));
const createCell = cell => ({ key: cell.key, children: cell.children, attrs: cell.attrs });
```

Cells are simple and just require a key, the children of said cell and then we use `attrs` to set `className` in our example like so.

```javascript
const attrs = { className: cx('content-display')};
```

> ## Creating a Cell Grid that looks like a Table

Reviewer: How confident are we in the accessibility of the cell grid like table?

Author: No longer applicable since CellGrid example has been removed.

    <div className={cx('cellgrid-wrapper')}>
      <CellGrid className={cx('row')}>
        <Cell className={cx('start-accessory')} key="cell-0" width={{ static: { value: 40, unit: 'px' } }}>{startAccessory}</Cell>
        <Cell className={cx('display')} key="cell-i" width={{ scalar: 1 }}>{display}</Cell>
        ...
      </CellGrid>
      ...
    </div>
  );
```

In this example, the `<CellGrid>` effectively act as rows and doesn't require any props other than defining a class for it. The `<Cell>`'s are where the content is placed. In the example provided at the bottom of this page, the content has two accessories on either side of the main information. Setting these accessories `width` prop to `{ static: { value: 40, unit: 'px' } }` allows the accessories to only take up a small amount of space, while the rest of the `Cell`'s use `{ scalar: 1 }` for width to automatically take up an equal percentage of the remaining space.

For the main styles it uses the following CSS.

```css
.cellgrid-wrapper {
  border-bottom: 1px solid #c8cacb;
}

.row {
  border-top: 1px solid #c8cacb;
  display: flex;
  min-height: 60px;
}

.row:nth-child(even) {
  background: #f4f4f4;
}
```

> ## Creating a read-only example using HTML table

Reviewer: I'd start with this example, the simplest, and work up the the most complex.

Author: No longer applicable since other examples have been removed.

If you are making a read-only table, we suggest using [HTML Table](/components/terra-html-table/html-table/table) which gets put together like this.

```javascript
const htmlTable = (
  <Table>
    <Header>
      <HeaderCell key="cell-date">{"Date"}</HeaderCell>
      <HeaderCell key="cell-1">{"Chrome"}</HeaderCell>
      ...
    </Header>
    <Body>
      <Row key="row_0">
        <Cell key="display_0" className={cx('cell-date')}>{"October 2019"}</Cell>
        <Cell key="display_1" className={cx('cell')}>{"68.91%"}</Cell>
        ...
      </Row>
      ...
    </Body>
  </Table>
);
```

It uses a straightforward Header and Body section. The header section is composed of `<HeaderCell>`'s which just needs a simple key and child for its display.

The body `<Row>`s and `<Cell>`s are also for the most part easily constructed with keys required for both. The `<Cell>`s also get classes for ease of styling and children for actual content.

## Example

> <ResponsiveListandTableExample title='Responsive List And Table Example' />

Reviewer: Intersperse these examples after each section where you talk about building them to break up the article.

Author: No longer applicable since other examples have been removed.
